test/test_foreach.py

```diff
@@ -77,23 +90,208 @@ def _get_test_data(self, device, dtype, N):
+            ]):
+                disable_fastpath = op.ref == torch.div and dtype in torch.testing.get_all_int_dtypes() + [torch.bool]
+                # int scalar
+                if i == 0:
+                    disable_fastpath |= dtype == torch.bool
+                # float scalar
+                if i == 1:
+                    disable_fastpath |= dtype in torch.testing.get_all_int_dtypes() + [torch.bool]
+                # bool scalar
+                if i == 2:
+                    disable_fastpath |= dtype == torch.bool
+                    if op.ref in (torch.add, torch.mul):
+                        disable_fastpath = False
+                # complex scalar
+                if i == 3:
+                    disable_fastpath |= dtype not in torch.testing.get_all_complex_dtypes()
+                print(scalar, disable_fastpath)
+                self._test_binary_op_scalar(device, dtype, op, N, scalar, True, disable_fastpath)
+
+    @dtypes(*torch.testing.get_all_dtypes())
+    @ops(foreach_binary_op_db)
+    def test_binary_op_scalar_slowpath(self, device, dtype, op):
+        scalar_values = [
+            random.randint(1, 10), 1.0 - random.random(), True, complex(1.0 - random.random(), 1.0 - random.random())]
+        for N in N_values:
+            for scalar in scalar_values:
+                self._test_binary_op_scalar(device, dtype, op, N, scalar, False, False)
+
+    def _test_binary_op_scalarlist(self, device, dtype, opinfo, N, scalarlist, is_fastpath, disable_fastpath):
+        n_expected_cudaLaunchKernels = N if disable_fastpath else 1
+        op, ref, inplace_op, inplace_ref = self._get_funcs(opinfo, n_expected_cudaLaunchKernels)
+        inputs = [opinfo.sample_inputs(device, dtype, N, noncontiguous=not is_fastpath), scalarlist]
+        self._regular_binary_test(dtype, op, ref, inputs, is_fastpath)
+        self._inplace_binary_test(dtype, inplace_op, inplace_ref, inputs, is_fastpath)
+
+    # note(mkozuki): Why two functions depending on with/without bool?
+    # `foreach_sub` & `foreach_sub_` do `sub_check(tensors[i], scalars[i])` from i=1...N.
+    # So, if scalarlist has one or more bool values, `foreach_sub` and `foreach_sub_`
+    # raise bool subtraction error before doing any math.
+    # While regular `sub` and `sub_` do some math until they encounter bool.
+    # So, foreach sub's throw bool sub error first. However, regular sub's throw different
+    # errors depending on the order of scalarlist. To keep actual unit test impl simple,
+    # separating mixed scalarlist tests. By setting the first element of scalarlist to bool,
+    # they are expected to throw bool sub error even in inplace test.
+    @skipCUDAIfRocm
+    @skipMeta
+    @ops(foreach_binary_op_db)
+    def test_binary_op_scalarlist_fastpath(self, device, dtype, op):
+        for N in N_values:
+            for i, scalarlist in enumerate([
+                [random.randint(0, 9) + 1 for _ in range(N)],
```

Comment:
instead of `enumerate` you can have a nested tuple there, with first element denoting type ("float", "int" or "mixed")